### DISCUSSION TITLE
Avoiding parallel threading in py-pde

### DISCUSSION THREAD

**Discussion by ArvinSubramaniam** [zwicker-group/py-pde#51] [Q&A]
Hi,

My question is regarding the use of parallel threading in py-pde.

From what I can tell, it turns out that py-pde is using all cores for simulations (see aattached photo of CPU usage on my linux computer - all 8 cores are used), although it is not clear that this is necessary (since no parallel computing is used). This drastically slows down simulations on my device, and also the computing cluster that I use.

![2021-02-03 16-19-39 யிலிருந்து திரைவெட்டு](https://user-images.githubusercontent.com/32795128/106791114-cac31280-6654-11eb-92e7-a18672f3d349.png)

An upon adding the following lines:

``` python
from pde.tools import
numba.NUMBA_PARALLEL = False
```

the problem still remains (see second photo - granted they look the same but trust me they were taken from before and after)

![2021-02-03 19-15-23 யிலிருந்து திரைவெட்டு](https://user-images.githubusercontent.com/32795128/106791289-0067fb80-6655-11eb-9d3d-64fa89f2365a.png)

The code is thus:

``` python
from pde import ScalarField,UnitGrid,PDE,FieldCollection,CartesianGrid,MemoryStorage,plot_kymograph,plot_kymographs

from pde.tools import *
numba.NUMBA_PARALLEL = False

n_grid = 256
t_start = time.time()
grid = CartesianGrid([[0,1],[0,1]],[n_grid,n_grid],periodic=True) ### Time step has to be smaller here

K1 = 2.1

K = K1/(n_grid**(2))

init_in = np.random.normal(0.3,0.2,(n_grid,n_grid))
init_ = ScalarField(grid,init_in)
init_.plot()

storage = MemoryStorage()
num_kym = 50
eq=PDE({"u":f"-{K}*laplace(laplace(u)) - laplace(u - u**(3))"})# define the pde
 

tend = 1
result2=eq.solve(init_, t_range=tend,tracker=[storage.tracker(tend/num_kym)])
result2.plot(title=r'$T={}$'.format(tend))

```

Is there a neat way enforce that a single core usage on the py-pde source code? Any help is much appreciated.

**Comment by david-zwicker** [comment]
I found a bug in `py-pde`, which prevented disabling multiprocessing. The current master branch should fix this. You should now be able to disable multiprocessing using the following construct
```python
from pde.tools import numba
numba.NUMBA_PARALLEL = False
```
Could you please test whether this fixes the problem?

**Reply by ArvinSubramaniam** [reply]
The problem is still unresolved. I've had a friend independantly verify this on his own device with the lastest py-pde. 

![2021-02-04 16-19-03 யிலிருந்து திரைவெட்டு](https://user-images.githubusercontent.com/32795128/106913920-cce4aa00-6704-11eb-90c3-aa2d8ccf6ee3.png)

**Reply by david-zwicker** [reply]
This is surprising, since it seems to work on my side. Could you please post the output of the following call after you set `numba.NUMBA_PARALLEL = False`:

```python
from pde import environment
environment()
```

**Reply by ArvinSubramaniam** [reply]
{'package version': '0.12.9',
 'python version': '3.8.5 (default, Sep  4 2020, 07:30:14) \n[GCC 7.3.0]',
 'platform': 'linux',
 'mandatory packages': {'matplotlib': '3.3.2',
  'numba': '0.51.2',
  'numpy': '1.19.2',
  'scipy': '1.5.2',
  'sympy': '1.7.1'},
 'matplotlib environment': {'backend': 'module://ipykernel.pylab.backend_inline',
  'plotting context': 'JupyterPlottingContext'},
 'optional packages': {'h5py': '2.10.0',
  'pandas': '1.2.1',
  'pyfftw': 'not available',
  'tqdm': '4.56.0'},
 'numba environment': {'version': '0.51.2',
  'parallel': False,
  'fastmath': True,
  'debug': False,
  'using_svml': False,
  'threading_layer': 'tbb',
  'omp_num_threads': None,
  'mkl_num_threads': None,
  'num_threads': 8,
  'num_threads_default': 8,
  'cuda_available': True,
  'roc_available': False}}

**Reply by david-zwicker** [reply]
Did you restart the kernel before setting `numba.NUMBA_PARALLEL = False`? If not, it could be that numba cached the result of the compilation and still used the old compiled version of the function.

If this does not help, you might be successful in [modifying the environment variable `NUMBA_NUM_THREADS`](https://numba.readthedocs.io/en/stable/reference/envvars.html?highlight=tbb#envvar-) to control the number of threads numba uses.

**Reply by ArvinSubramaniam** [reply]
Have imposed the above. I've set NUMBA_PARALLEL=False in my local numba.py file. In addition, one can impose NUMBA_NUM_THREADS=x python3 programname.py where x is the numer of threads to enforce the number explicitly. This will lead to environments of the following form:

{'version': '0.51.2', 'parallel': False, 'fastmath': True, 'debug': False, 'using_svml': False, 'threading_layer': 'tbb', 'omp_num_threads': None, 'mkl_num_threads': None, 'num_threads': 1, 'num_threads_default': 8, 'cuda_available': True, 'roc_available': False}

Problem is, when running even simple code (i.e the diffusion example for instance), all 8 cores are still used. So there may be something in the PDEBase/Controller objects that are imposing all 8 cores to be used. If I run a dummy for loop inside the Controller.run function for instance (i.e whenever eq.solve is called this line of code will run), I see that only one core is used. This can be seen in the image below. When eq.solve the first dummy for loop is run, only core is used (Linux shuffles it around), but when the subsequent make_stepper, etc. are called all cores are used.

![2021-02-04 23-22-21 யிலிருந்து திரைவெட்டு](https://user-images.githubusercontent.com/32795128/106963705-4601f280-6741-11eb-9e3d-c72527556a8b.png)


So it seems that there is something in the subsequent function being called that imposes this parallelism. I will dig deeper tomorrow.

**Reply by david-zwicker** [reply]
Thanks for investigating this further. I think the culprit might be the function [`scipy.intergrate.solve_ivp`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html), which is used for time-stepping when no fixed time step is supplied. I don't know how to control multi-processing in `scipy`, but you can switch to a Numba-compiled explicit time stepper by supplying a fixed time step using the [`dt` argument in the call to `eq.solve`](https://py-pde.readthedocs.io/en/latest/packages/pde.pdes.html#pde.pdes.base.PDEBase.solve).

**Reply by ArvinSubramaniam** [reply]
Thanks for this. Indeed this helps.

Another (potneitally interesting finding) is that, even with ``` dt``` specified, the parallelism seems to depend on the exact grid size, `L`. I've inlcuded belwo CPU usage for `L=213` vs `L=256`. The former uses one core whereas the latter has heavy parallelism. Interestingly, paraellism for larger `L=1024` is actually less than 256, for reasons I haven't figured out.

L=213
![modelB_dt_213](https://user-images.githubusercontent.com/32795128/107061337-57e2a480-67d8-11eb-8035-cb99b74e2a5c.png)

L=256
![modelB_dt_256](https://user-images.githubusercontent.com/32795128/107061352-5dd88580-67d8-11eb-841f-3e2f9fcdc0a0.png)

So to conclude, the best bet seems to be modest grid sizes `L<256` that are not powers of 2.

**Reply by david-zwicker** [reply]
Yes, `py-pde` disables parallelism when evaluating differential operators on small grids. To test whether the multiprocessing is actually disabled, you thus need to look at grids that are large enough.

**Reply by david-zwicker** [reply]
Just as a quick follow-up. I changed the way settings are handled in the latest version of `py-pde`, so multi-processing should now be disabled using the following construct:

```python
from pde import config
config['numba.parallel'] = False
```

Note that the old way might still work for about half a year, but it will be removed eventually.

**Reply by ArvinSubramaniam** [reply]
Thanks!
